rfcs/20200205-standalone-keras-repository.md

```diff
@@ -0,0 +1,351 @@
+hours to be reflected on the Keras side.
+* The Keras code will be mirrored to a Google-internal code repository via Google-internal 
+tools within a very short time window after each change.
+The Google-internal CI tests will run on HEAD for both Keras and TF code.
+* The CI build for the repository on GitHub might break when it sees a
+new version of `tf-nightly`, if  certain behavior has been changed and wasn't
+caught by unit tests. We have  observed a few similar cases with
+[tf/addons](https://github.com/tensorflow/addons).
+We hope this can be reduced by stronger
+unit test coverage by Google internel systems, when both TF and Keras code are 
+tested at HEAD.
+* pip package management. Keras will now follow the `tf-estimator` approach. 
+"pip install tensorflow" should also install Keras (from PyPI) as well.
+There are more details for the pip package in the
+[Improved pip package structure](https://github.com/tensorflow/community/pull/182) RFC.
+
+### Dependency Cleanup
+
+As the high-level API of TensorFlow, Keras should have a direct dependency on
+TF low-level APIs, but not the other way around. Unfortunately, there is some existing reverse 
+logic in the TF code that relies on Keras, which we should update/remove 
+when we split the repository.
+
+The current usage of Keras from TensorFlow are:
+* Unit tests, which should be converted to integration tests, or port the tests
+to Keras repository.
+* `feature_column`.
+* Legacy `tf.layers` in v1 API.
+* legacy RNN cells.
+* TPU support code for `optimizer_v2`.
+* SavedModel.
+* TF Lite.
+
+All Keras imports in integration tests can be changed to use dynamic import like below:
+
+```python
+try:
+   from tensorflow.python.keras.engine import base_layer
+except ImportError:
+   tf.logging.error('keras is not installed, please pip install keras')
+   base_layer = None
+```
+
+### Update Keras to only use public TF APIs
+
+The current Keras code will still work if we do e.g.:
+```python
+from tensorflow.python.ops import array_ops
+
+ones = array_ops.ones([2, 3])
+```
+
+However, since Keras is a separate repository, having it only use TF 
+public APIs will heavily reduce the chance of breakage caused by relying
+on private methods or implementation details. We think this point is
+critial to the health of the project. This also allows TF to change internal 
+implementation details without worrying about breaking Keras.
+
+The converted code should look like e.g.:
+
+```python
+import tensorflow as tf
+
+ones = tf.ones([2, 3])
+```
+
+During this conversion, we might notice that certain TF features used in Keras are
+not public. A decision should be made on a case-by-case basis:
+
+* Copy the functionality from TF to Keras.
+* Replace the usage with another alternative TF public API.
+* Make the functionality a new TF public API.
+
+**Note that the open-source community is encouraged to contribute to this effort.**
+
+### Two-stage change process
```

Comment:
Thanks! Great description of the biggest problems we may encounter.

I would like us to consider targeting tensorflow stable (or even release candidates). This is the move we made in Tensorflow addons.

Some of the points to support this idea:

* As said in the meeting, changes happening in both tensorflow and keras at the same time are very rare.
* The second scenario should not happened as said in this RFC (or we should work to avoid it), because we shouldn't rely on private implementations of tensorflow.
* Pinning the version number of dependencies in the CI is considered a best practice.
* Sudden breaking/bugs in the TF side can make our build fail for a undertimined number of days (until we find the root cause) and during this period of time we can't merge any pull requests.
* For local testing, developers need to download tensorflow nightly every day.
* The build is not deterministic across time (I can't reproduce easily the tests from 8 days ago let's say)
* Since the tests are not deterministic, we can't use tools like git biscect to catch the commit which introduced a bug in the codebase.

See the move we made for tensorflow addons:
https://github.com/tensorflow/addons/pull/912

Since I understand that setting in stone the rule of targeting TF stable is very challenging for rapid developement (scenario 1 and 2), I propose a solution in between:

* We can target the stable or the nightly version, but the version must always be pinned. e.g. in the CI we have `pip install tensorflow==2.1.0` or `pip install tf-nightly==2.2.0.dev20200207`.
* When scenario 1 and 2 presented in this RFC are not happening, we target the stable version: `pip install tensorflow==2.1.0`

If scenario 1 or 2 happens:

1) Keras is targeting `tensorflow==2.1.0`
2) User makes the PR 22222 to tensorflow
3) PR 22222 gets merged
4) tf-nightly 2.2.0.dev20200207 gets released
5) PR 33333 is made to keras. We need to be able to merge this feature quickly.
6) The PR exceptionally change the pinned version of tensorflow targeted in the CI `tensorflow==2.1.0` ->  `tf-nightly==2.2.0.dev20200207`
7) The PR is merged.
8) Tensorflow 2.2.0 is out.
9) We make a PR to change the pinned version in the CI `tf-nightly==2.2.0.dev20200207` -> `tensorflow==2.2.0`.

With this, we have the best of both worlds. We target a stable version most of the time, and even when we don't, the version is pinned so the tests are deterministic.

Reply:
Thanks @gabrieldemarmiesse for the detailed suggestion. I do like the idea of pinning to a certain version which should greatly reduce the chance of breakage. I think pinning the version (nightly or stable) is a very good alternative for reproducibility and supporting the rapid development.

Reply:
Updated with 86d2c37. I will leave this open since your comment contains more details.

Reply:
My comment has been adressed. Could you minimize this thread for readability? Thanks!